### ISSUE TITLE
ubsan load/store misaligned address errors from memory_stream_test and gguf_parser_test

### ISSUE THREAD

**Issue by yuennancy** [iree-org/iree#16608]
### What happened?

Was running asan tests memory_stream_test and gguf_parser_test downstream, and got illegal instructions errors.  Turned on ubsan and got more detailed errors.

Then tried it with a copy of github/openxla/iree @ head, compiled with ubsan.  ctest don't fail but the LastLog shows the runtime error messages.  Also see it in raw test output when test binaries are run directly.

[ RUN      ] MemoryStreamTest.Fill
/opt/work/iree/github-head/openxla/iree/runtime/src/iree/io/memory_stream.c:267:9: runtime error: store to misaligned address 0x7ffdba454351 for type 'uint16_t' (aka 'unsigned short'), which requires 2 byte alignment
0x7ffdba454351: note: pointer points here
 56 00 00  dd dd dd dd dd dd dd dd  dd dd dd dd dd dd dd dd  50 81 45 8d 54 56 00 00  a8 53 45 ba fd
              ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /opt/work/iree/github-head/openxla/iree/runtime/src/iree/io/memory_stream.c:267:9 in 

[ RUN      ] GgufFormatTest.SingleTensor
/opt/work/iree/github-head/openxla/iree/runtime/src/iree/io/formats/gguf/gguf_parser.c:417:22: runtime error: load of misaligned address 0x558d6855052e for type 'const uint64_t' (aka 'const unsigned long'), which requires 8 byte alignment
0x558d6855052e: note: pointer points here
 02 00 00 00 02 00  00 00 00 00 00 00 02 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00
             ^ 
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /opt/work/iree/github-head/openxla/iree/runtime/src/iree/io/formats/gguf/gguf_parser.c:417:22 in 

More detailed output at memory_stream_test: https://gist.github.com/yuennancy/e0ad695cc2d5097ccbd4a16872d13382
and gguf_parser_test: https://gist.github.com/yuennancy/77d268f08721b5dc72fe3bce8f7525ae

### Steps to reproduce your issue

1. Configure cmake with ubsan:
```cmake -G Ninja -B ../iree-build/ -S .     -DCMAKE_BUILD_TYPE=Debug     -DIREE_ENABLE_ASSERTIONS=ON     -DIREE_ENABLE_SPLIT_DWARF=ON     -DIREE_ENABLE_THIN_ARCHIVES=ON     -DCMAKE_C_COMPILER=clang     -DCMAKE_CXX_COMPILER=clang++     -DIREE_ENABLE_LLD=ON -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache -DIREE_ENABLE_UBSAN=ON```

2. Build memory_stream_test and gguf_parser_test:
```cmake --build ../iree-build/ --target runtime/src/iree/io/memory_stream_test --target runtime/src/iree/io/formats/gguf/gguf_parser_test```

3. Run ctest and view Lastlog output.  ctest passes, but errors will show up in Lastlog:
```ctest --test-dir ../iree-build/  -R memory_stream_test; cat ../iree-build/Testing/Temporary/LastTest.log```

```ctest --test-dir ../iree-build/  -R gguf_parser_test; cat ../iree-build/Testing/Temporary/LastTest.log```

4. Run tests directly.  Tests pass, but errors are in the test output:
```../iree-build/runtime/src/iree/io/memory_stream_test```

```../iree-build/runtime/src/iree/io/gguf_parser_test```

### What component(s) does this issue relate to?

Runtime

### Version information

Checked out IREE @ head

**Comment by bjacob** [comment]
@benvanik, The first report about this line, https://github.com/openxla/iree/blob/76cbaaca2cf44ac0e4f58cbce4134c253e38f758/runtime/src/iree/io/memory_stream.c#L267

And the code path for loading an unaligned uint16 goes here:
https://github.com/openxla/iree/blob/76cbaaca2cf44ac0e4f58cbce4134c253e38f758/runtime/src/iree/base/alignment.h#L364-L366

It is always undefined behavior to have a C or C++ object at a memory location whose address isn't a multiple of its alignment, so the UBSan report seems to be correct. No platform-specific trait can change that. Instead, platform-specific traits such as `__ARM_FEATURE_UNALIGNED` only merely mean an existence claim, that there is *some* way to perform an unaligned access. They don't suddenly change the C/C++ semantics to allow unaligned objects. In the case of `__ARM_FEATURE_UNALIGNED`, the Arm reference https://developer.arm.com/documentation/dui0774/g/chr1383660321827 explains __ARM_FEATURE_UNALIGNED as "Set to 1 if the target supports unaligned access in hardware".  This could mean a number of things (I don't know exactly in this case), such as possibly meaning that some builtin or intrinsic is available to perform that, I dont know --- but `alignment.h` in its current form is undefined behavior and the easy way out for now would be to just build with `IREE_MEMORY_ACCESS_ALIGNMENT_REQUIRED_*`.

**Comment by yuennancy** [comment]
Platform I ran on is x86_64.

So when using the unaligned functions, it's behaving as intended and the illegal instruction error I got from the test can be safely ignored (thanks)

What about the error on https://github.com/openxla/iree/blob/main/runtime/src/iree/io/formats/gguf/gguf_parser.c#L417?

[ RUN ] GgufFormatTest.SingleTensor
/opt/work/iree/github-head/openxla/iree/runtime/src/iree/io/formats/gguf/gguf_parser.c:417:22: runtime error: load of misaligned address 0x558d6855052e for type 'const uint64_t' (aka 'const unsigned long'), which requires 8 byte alignment
0x558d6855052e: note: pointer points here
02 00 00 00 02 00 00 00 00 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00

**Comment by bjacob** [comment]
The gguf_parser.c one is less obvious for me to read. Hopefully @benvanik can pick up from here.

**Comment by julianwa** [comment]
Unrelated to whether we should act on the errors: Is it expected behavior that these errors don't fail the tests?

**Comment by bjacob** [comment]
yes, typical. this undefined behavior could translate into anything in theory, but when the target load/store instructions don't have any alignment requirement, that would "just work". When the target load/store instructions do have an alignment requirement, the observed behavior would be a crash. On some target architectures, both kinds of instructions exist so it can be a matter of luck that determines in which cases you fall. And when the compiler can tell at compile time that the address is going to be misaligned, some wilder things can happen, such as, simply omitting the load/store on the basis that "it would be undefined behavior anyway if it happened".
